### PR TITLE
[modbus] Modbus register array backed by bytes and other simplifications

### PR DESCRIPTION
**Pull request by ssalonen**
Inspired by #8172 I reworked the modbus transport to use direct byte arrays to represent the data. 

This is purely refactoring with the aim of reducing boilerplate byte manipulation/copying code in bindings. With the changes we do get boosted performance: with 45000 invocations to `extractStateFromRegisters` with varying inputs, the self time is dropped to 55% of original, total time to 80%. Naturally these type of performance gains are dwarfed by network I/O and other real time overheads.

This has many benefits:
- Unnecessary boxing / unboxing of ModbusRegister objects is removed all-together and one can operate directly with familiar `byte[]` using bitwise operations
- Some modbus bindings constructed `byte[]` from `ModbusRegister[]` (essentially array of 16-bit data blocks, registers). 

In addition I brought in more general methods to work with the data:
- Methods such as `extractXX`, e.g. `extractSInt8` (extract 8-bit signed integer) working directly with `byte[]` data. Return type is directly smallest suitable java primitive such as `byte`, `short` etc.
- `ValueBuffer` interface that acts similar to java's `ByteBuffer` but having specialized methods to extract different types of value types that are encountered with modbus

Finally, simplifications & harmonization was made to how registers are converted to to numbers, and how commands are converted to bytes. I find the new conversion logic more readable.

The PR is splitted in two parts:
1. changes in transport API. Mainly backwards compatible. Only exception is extractStringFromXX which now returns `String`, not `StringType`
2. Updates and simplifications to individual bindings. Tests are added, if missing, to have regression tests for data type conversions

### CODE REVIEW

**Comment by ssalonen** [comment]
CC code owners to review at least the changes in the relevant binding

/bundles/org.openhab.binding.modbus.e3dc/ @weymann
/bundles/org.openhab.binding.modbus.helioseasycontrols/ @bern77
/bundles/org.openhab.binding.modbus.stiebeleltron/ @pail23
/bundles/org.openhab.binding.modbus.studer/ @giovannimirulla
/bundles/org.openhab.binding.modbus.sunspec/ @mrbig

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusRegisterArray.java [hunk 1]
```diff
@@ -12,65 +12,70 @@
  */
 package org.openhab.io.transport.modbus;
 
-import java.util.Iterator;
-import java.util.stream.IntStream;
-
+import org.bouncycastle.util.Arrays;
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusRegisterArray.java:15]: Wrong import, causes the build failure

**Comment by fwolter** [comment]
PRs must be filed against the main branch. See #8512. Can you create another PR and close this?

**Comment by ssalonen** [comment]
Thanks @fwolter . Was there a bot to automate this?

**Comment by ssalonen** [comment]
Closing in favor of #8865

**Pull request closed without merging**